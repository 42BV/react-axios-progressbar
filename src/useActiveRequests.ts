import { useEffect, useState } from 'react';
import { AxiosInstance, AxiosResponse, InternalAxiosRequestConfig } from 'axios';

export function useActiveRequests(axiosInstance: AxiosInstance) {
  const [ activeRequests, setActiveRequests ] = useState(0);

  function onNewRequest(config: InternalAxiosRequestConfig | Promise<InternalAxiosRequestConfig>) {
    setActiveRequests(activeRequests + 1);
    return config;
  }

  function decreaseActiveRequests() {
    setActiveRequests(Math.max(activeRequests - 1, 0));
  }

  function onRequestFulfilled(response: AxiosResponse) {
    decreaseActiveRequests();
    return response;
  }

  function onRequestRejected(error: unknown) {
    decreaseActiveRequests();
    return Promise.reject(error);
  }

  useEffect(() => {
    const requestInterceptor = axiosInstance.interceptors.request.use(onNewRequest);
    const responseInterceptor = axiosInstance.interceptors.response.use(onRequestFulfilled, onRequestRejected);

    return () => {
      axiosInstance.interceptors.request.eject(requestInterceptor);
      axiosInstance.interceptors.response.eject(responseInterceptor);
    }
  }, [axiosInstance]);

  return activeRequests;
}
