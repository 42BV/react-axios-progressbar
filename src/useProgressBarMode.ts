
/*
  The modes form a state machine with the following flow

  init -----> active -> complete --
  ^  \                            |
  |   \----\                      |
  |         \                     |
  |          \                    |
  |           â–¾                   |
  ---------  hibernate <-----------

  hibernate: no animation is running the bar is invisible
  init: Preparing to potentially show the animation.
  active: the animation is running slowly to 80%
  complete: the animation runs quickly to 100%
*/
import { useEffect, useState } from 'react';

export type Mode = 'hibernate' | 'init' | 'active' | 'complete';

export function useProgressBarMode() {
  const [ mode, setMode ] = useState<Mode>('hibernate');
  const [ activeRequests, setActiveRequests ] = useState(0);

  useEffect(() => {
    if (mode === 'complete') {
      // Moving to hibernate after 1 second to allow the close animation to complete
      const timeout = setTimeout(() => {
        setMode('hibernate');
      }, 1000);

      return () => {
        clearTimeout(timeout);
      };
    }

    if (mode === 'active' && activeRequests === 0) {
      // No pending requests anymore, move to complete if there are still no pending requests after 200 milliseconds
      const timeout = setTimeout(() => {
        setMode('complete');
      }, 200);

      return () => {
        clearTimeout(timeout);
      };
    }

    if (mode === 'init') {
      if (activeRequests > 0) {
        // There are pending requests, move to active if there are still pending requests after 100 milliseconds
        const timeout = setTimeout(() => {
          setMode('active');
        }, 100);

        return () => {
          clearTimeout(timeout);
        };
      } else {
        // No pending requests anymore, move to hibernate
        setMode('hibernate');
      }
    }

    if (activeRequests > 0 && mode === 'hibernate') {
      setMode('init');
    }
  }, [ activeRequests, mode, setMode ]);

  function newRequest() {
    setActiveRequests(activeRequests + 1);
  }

  function requestCompleted() {
    setActiveRequests(activeRequests - 1);
  }

  return { mode, newRequest, requestCompleted };
}
