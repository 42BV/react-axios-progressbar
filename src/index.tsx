import React from 'react';
import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import { useProgressBarMode } from './useProgressBarMode';

type Props = {
  style?: Record<string, unknown>;
};

export function ProgressBar({ style }: Props) {
  const { mode } = useProgressBarMode();

  if (mode === 'hibernate') {
    return null;
  }

  const width = mode === 'complete' ? 100 : mode === 'init' ? 0 : 80;
  const animationSpeed = mode === 'complete' ? 0.8 : 30;

  return (
    <div
      className="react-axios-progress-bar"
      style={{
        position: 'absolute',
        top: '0',
        zIndex: 9000,
        backgroundColor: '#f0ad4e',
        height: '4px',
        ...style,
        width: `${width}%`,
        transition: `width ${animationSpeed}s ease-in`
      }}
    />
  );
}

export function onNewRequest(config: AxiosRequestConfig) {
  const { newRequest } = useProgressBarMode();
  newRequest();
  return config;
}

axios.interceptors.request.use(onNewRequest);

export function onRequestFulfilled(response: AxiosResponse) {
  const { requestCompleted } = useProgressBarMode();
  requestCompleted();
  return response;
}

export function onRequestRejected(error: unknown) {
  const { requestCompleted } = useProgressBarMode();
  requestCompleted();
  return Promise.reject(error);
}

axios.interceptors.response.use(onRequestFulfilled, onRequestRejected);
