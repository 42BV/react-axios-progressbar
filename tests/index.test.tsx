import React from 'react';
import { ProgressBar } from '../src/index';

import * as ProgressBarMode from '../src/useProgressBarMode';
import { Mode } from '../src/useProgressBarMode';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Component: ProgressBar', () => {
  function setup({ mode, style }: { mode: Mode; style?: Record<string, unknown> }) {
    jest.spyOn(ProgressBarMode, 'useProgressBarMode').mockReturnValue({
      mode
    });

    const { container } = render(<ProgressBar style={style} />);

    return { container };
  }

  describe('ui', () => {
    test('mode: hibernate', () => {
      const { container } = setup({ mode: 'hibernate', style: undefined });
      expect(container.firstChild).toBeNull();
    });

    test('mode: init', () => {
      const { container } = setup({ mode: 'init', style: undefined });
      expect(container).toMatchSnapshot();
      expect(container.firstChild).toHaveStyle({ width: '0%', transition: 'width 30s ease-in' });
    });

    test('mode: active', () => {
      const { container } = setup({ mode: 'active', style: undefined });
      expect(container.firstChild).toHaveStyle({ width: '80%', transition: 'width 30s ease-in' });
    });

    test('mode: complete', () => {
      const { container } = setup({ mode: 'complete', style: undefined });
      expect(container.firstChild).toHaveStyle({ width: '100%', transition: 'width 0.8s ease-in' });
    });

    test('custom style', () => {
      const { container } = setup({
        mode: 'active',
        style: { backgroundColor: 'red', height: '10px', left: '10px' }
      });
      expect(container.firstChild).toHaveStyle({ backgroundColor: 'red', height: '10px', left: '10px' });
    });
  });
});
