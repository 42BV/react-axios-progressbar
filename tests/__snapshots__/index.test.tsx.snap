// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ProgressBar ui mode: init 1`] = `
<div>
  <div
    class="react-axios-progress-bar"
    style="position: absolute; top: 0px; z-index: 9000; background-color: rgb(240, 173, 78); height: 4px; width: 0%; transition: width 30s ease-in;"
  />
</div>
`;
